### DISCUSSION TITLE
Working with Library Panels - missing details

### DISCUSSION THREAD

**Discussion by techuser12345** [esnet/gdg#286] [General]
**Describe the bug**
I use GDG to export the dashboard and push dashboard *.json files to stash under: grafana/dashboards/folders/*.json
I put those *.json files in prod as grafana provisioning. I also have datasource.yml
I notice "Panel plugin not found: Unable to load library panel: XXXXXX-XXXXXX-XXXXX" error in prod.

I went through this document https://software.es.net/gdg/docs/tutorials/working-with-library-panels/
You have listed rules. But the solution is not clear.
Are you suggesting this is the workaround/solution we should use 'The only way I can see to move a lib element is to unlink the panel, delete the panel and re-create it in a different folder, then re-link it.'
If so the steps mentioned below or the order do not match the statement.

I would like to know how to make library panel work in prod environment using provisioning and gdg is there a way where we can make it available please?

**To Reproduce**
1. in Grafana UAT GUI create dashboards, library panel.
2. Run GDG to export dashboards to stash
3. set up grafana provisioning so the dashboards are put in a path so grafana provisioning can pick up
4. load a dashboard with library panel you will see error ""Panel plugin not found: Unable to load library panel: XXXXXX-XXXXXX-XXXXX" "
5. UAT will not have error as the library panels were created in Grafana UAT GUI.

**Expected behavior**
Step-by-step instruction how to export library panels from UAT and make them available in prod - achievable via automation (bash script) not manual GUI steps ,as there will be several library panels linked to different dashboards (one to many relationships).

**Desktop (please complete the following information):**
 - GDG Version: 0.4.1
 - Grafana Version: v9.5.2 
 - Configuration for Importer

**Additional context**

**Comment by safaci2000** [comment]
Library panels are just weird in Grafana.  I went down the rabbit hole when I was working on the feature, but let me see if I can remember.  

The main thing is that you can't import a dashboard into Grafana prior to the library component existing.  So once you have a library components in your staging ENV, you pull it down, pull the associated dashboards and then export them out to production.

You just can't upload a dashboard while the library component doesn't exist, and you can't delete a library component while a dashboard references it.  

I have a test for that, so my first step is:

```sh
gdg b libraryelements upload 
```

out:
```
2024-08-15 18:43:33 INF Reading files from folder folder=test/data/org_main-org/libraryelements
┌────────────────────────────────────────────────────────────────┐
│ NAME                                                           │
├────────────────────────────────────────────────────────────────┤
│ Dashboard Makeover - Extra Cleaning Duty Assignment Today      │
│ Dashboard Makeover - Lighting Status                           │
│ Dashboard Makeover - Side Dish Prep Times, past 7 days         │
│ Dashboard Makeover - Time since we purchased these spices      │
│ Extreme Dashboard Makeover - Grill                             │
│ Extreme Dashboard Makeover - Mac Oven                          │
│ Extreme Dashboard Makeover - Refrigerator Temperature (F)      │
│ Extreme Dashboard Makeover - Room Temperature (F)              │
│ Extreme Dashboard Makeover - Salmon Cooking Times, past 7 days │
└────────────────────────────────────────────────────────────────┘
```

Then you can List the dashboards that have a connection with the given panel ID:  (Latest release has a bug with this though I can push a small patch soon ).

```
./bin/gdg b libraryelements list-connections T47RSwQnz
```

The T47...etc is the library panel ID.

```sh
2024-08-15 18:52:09 INF Running Sanity Check of Organization Membership
2024-08-15 18:52:09 INF Listing library connections for context context=testing
┌────┬───────────┬──────────────────────────────┬──────────────────────────────┬────────┐
│ ID │ UID       │ SLUG                         │ TITLE                        │ FOLDER │
├────┼───────────┼──────────────────────────────┼──────────────────────────────┼────────┤
│ 12 │ F3eInwQ7z │ dashboard-makeover-challenge │ Dashboard Makeover Challenge │ Other  │
└────┴───────────┴──────────────────────────────┴──────────────────────────────┴────────┘

```

If you look at the JSON for the dashboard you'll see that the UID for the libarary panel matches:

```json
      "libraryPanel": {
        "description": "",
        "meta": {
          "connectedDashboards": 3,
          "created": "2022-04-27T21:27:23Z",
          "createdBy": {
            "avatarUrl": "/avatar/579fc54abdc9ab34fb4865322f2870a1",
            "id": 13,
            "name": "mike.johnson@grafana.com"
          },
          "folderName": "mj",
          "folderUid": "R0bMCcW7z",
          "updated": "2022-04-27T21:29:50Z",
          "updatedBy": {
            "avatarUrl": "/avatar/579fc54abdc9ab34fb4865322f2870a1",
            "id": 13,
            "name": "mike.johnson@grafana.com"
          }
        },
        "name": "Dashboard Makeover - Extra Cleaning Duty Assignment Today",
        "type": "table",
        "uid": "T47RSwQnz",
        "version": 2
      },
```

---

To create this in production, you just need to.

1. Upload Library panels to your production instance and upload your dashboards.  The IDs have to match but otherwise this should be pretty seamless.

**Comment by safaci2000** [comment]
https://github.com/esnet/gdg/pull/283 Fixes the listing issue.

**Reply by techuser12345** [reply]
Thanks.
I ran
./gdg backup libraryelements download
I have about 30 libraries
I am looking for a solution that I can include as part of the bash script I have in place that downloads all dashboards from uat to stash which is then pushed to prod.
The above solution seems too manual ?
Also I tried to find the uid from the json path I got as a result from ./gdg backup libraryelements download

grep uid /path/to/library/json
I got multiple uid result I tried to
./gdg backup libraryelements list-connections <uid>

Unable to retrieve a valid connection for <uid>
I tried looking for multiple json uid same error
Is there a way where I can deal with the entirely as a bulk push from uat to prod like you have arranged for dashboards please?
Please note I have latest gdg  0.7.2 and grafana on 11.2.0

Please advise as it’s really important. We are missing library panels in prod but the ones created in uat are there but we are unable to push them to prod as it’s tricky

**Reply by safaci2000** [reply]
Hey @techuser12345, sorry I thought I had responded to this but it seems I've missed it somehow. 

You should not need to mess with UID as long as you have the same dashboards on both systems. 

If you need to get the UID, simply use jq. 

cat libraryelements/General/foobar.yml | jq ".uid"

My suggestion is to pull the dashboard AND libraryelements on a grafana instance. Get everything the way you like it. Then download libraryelements and your dashboards. Once you have that, point to prod and replace the date with the working version

**Reply by techuser12345** [reply]
Thanks for your reply, When I run 
./gdg backup libraryelements download
I can download library panels in JSON format.
Where can I place these JSON files in Grafana. ie how to actually import those library panel jsons back in prod grafana? should i use gdg, api, sqlite db?

**Reply by safaci2000** [reply]
You would have two contexts pointing to the same export location and simply use gdg to push to prod.

**Reply by techuser12345** [reply]
Thanks for your response. I’d like to clarify the setup I mentioned in the initial ticket:

Dashboards and library panels are created manually in Grafana UAT GUI.

I run GDG in UAT to download dashboards only, and push them to stash.

In production, I use Grafana provisioning to load dashboards and datasources on startup — dashboards are placed at grafana/dashboards/folders/dashboards.json, and datasources are provisioned via deleteDatasources: and datasources: configs.

When loading a dashboard that references a library panel, I see the error:
"Panel plugin not found: Unable to load library panel: <UID>"
This does not occur in UAT, as the library panels exist there (created manually in the GUI).

Currently GDG is only set up in UAT. If I enable GDG in prod, with proper context for both UAT and PROD, and export/import between them — how would this interact with Grafana provisioning?

Also, aside from dashboards, what else should I export from UAT and import into PROD using GDG (e.g., library panels, folders, etc.) to prevent such errors?

As this is a production system, I want to be extra cautious and avoid any unintended impact. Appreciate your guidance.

**Reply by safaci2000** [reply]
At this point it's essentially a matter of opinions. The whole point of setting up an ENV like UAT that I usually call staging is to ensure that whatever changes you have will be pushed to production and work as expected. 

So it doesn't really matter what you use. Use GDG or provisioning... but you need to have an Env no matter what you call it that reproduces the steps that prod takes. It needs to load data the same way.

What you're describing is:

1. You have a UAT env where users create dashboards.
2. You pull from UAT to a new pattern that will now load using `Grafana provisioning`. 

What you don't have is an environment where you can test the grafana provisioning will work that mimics what prod does. So even if your test env is a docker compose stack, that's still better than simply taking UAT data and shipping to prod and hope it works. 

UAT, staging, test, qa whatever you call it, the env needs to load data the same exact way as prod. Be on the same version. It's supposed to be a clone of prod that nobody will care about. Then run your "release" on you test env to ensure it works correctly. 

The JSON data that GDG pulls are mostly what the API supports, but there's no guarantee to be the case.  The export step has logic that GDG performs. If you don't rely on GDG to do the export then I'm not really sure what the behavior will be. 

There's even entities where I created my own data format for my use case that grafana would have no idea what to do with. You just lucked out that the dash is fairly simple. 

For example, GDG opted to use the /import endpoint which support library elements instead of: `/api/dashboards/db` which does not. 

Pulling the JSON with one tool and handing it off to another tool will almost guarantee to give you trouble. 

Either ways.. I think the point I'm trying to get to is be consistent on the ENV you're testing this behavior on and how you're loading it. They need to use the exact pattern.

-----

That being said, I don't use Grafana provisioning.  I have two contexts setup, one for prod and one for staging. The Release does everything related to grafana except the dashboard provisioning. If something new was to be released, I simply re-run `gdg backup dash upload` or limit it by folder/dash-uid so I'm only uploading the item I'm interested in. 

I also test the "upload" on UAT using GDG to ensure it can re-write it correctly, including deleting/re-creating the lib panels.  You might still run into issued but I at least repeated the steps from the beginning...deleted add dash, re-uploaded then and saw they all looked good, so doing the same on prod would be fairly safe since the pattern has been tested.
